docs/CompilationExecution.md

```diff
@@ -0,0 +1,121 @@
+
+In a nutshell, the process is separated in compilation and execution. The Flux compiler is responsible to turn the initial
+representation of a query into an Intermediate Representation (IR) that makes sense for the execution step.
+The Flux Virtual Machine ([FVM](VirtualMachine.md)) is in charge of interpreting the IR and execute it to output its results.
+In figure:
+
+![chain](compilation_execution.png)
+
+ - The initial representation of the query can be either a raw string or an already parsed Abstract Syntax Tree.
+ - The compilation process chains multiple compilation steps.
+ `Compiler`s turn a query representation in a language into another one, until they reach the required IR.
+ - The `Interpreter` takes the IR and interprets it.
+ Some operations are trivial ---e.g., `1 + 2` or `obj.b` or `arr[0]`--- and do not require any further step to be executed.
+ For data manipulation pipelines, instead, the `Intepreter` needs more.
+ - The `Interpreter` does not know how to execute data pipelines.
+ So, it takes their representation ---i.e., the `Spec`-- and optimizes it by applying subsequent _transformations_,
+ in order to delegates execution to the `Engine`.
+ Those transformations are logical and physical planning. We consider these transformations as further compilation
+ steps, in that they turn a query representation into another.
+ - Finally, the `Engine` executes the optimized representation and outputs the results.
+
+Wrapping up, the actors that play in this process are `Compiler`s ---that turn a representation in some language into another one---
+and `Executor`s ---that execute a representation and produce results.
+Both the `Interpreter` and the `Engine` are `Executor`s:
+ - The `Engine` actually executes data manipulation pipelines and produces results;
+ - The `Interpreter` directly executes trivial operations and delegates pipeline execution to the `Engine`.
+ In its essence, it executes a higher level representation for the query than the `Engine`'s.
+ 
+In order to carry out execution, `Executor` need to know some information on a per-query basis
+(for instance, its memory limits, the request context, etc.).
+This information is embedded into the `ExecutionContext`.
+ 
+These are the interfaces:
+
+```go
+// Language is a language used to express a Flux query.
+type Language interface {
+    LanguageName() String
+}
+
+// Representation contains the actual content of a Flux query expressed in some Language.
+type Representation interface {
+    Lang() Language
+}
+
+// Results are the results of a Query.
+type Results interface {}
+// Statistics are the statistics of a Query execution.
+type Statistics interface {}
+
+// Query is an executed Flux query.
+// It provides its representation, results, error, and statistics.
+type Query interface {
+    Results() Results
+    Error() error
+    Stats() Statistics
+    Representation() Representation
```

Comment:
Why does a query need to know its representation?

Reply:
I thought about that for debugging. If you want you can turn a `Representation` into a string (somehow) and log it. I can just remove that and thought about it later.